source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,667 @@
+    case SpvOp::SpvOpSpecConstantComposite:
+      break;
+    default:
+      return nullptr;
+  }
+  return CreateConst(GetType(inst), literal_words_or_ids);
+}
+
+Const* FoldSpecConstantOpAndCompositePass::FindRecordedConst(uint32_t id) {
+  auto iter = id_to_const_val_.find(id);
+  if (iter == id_to_const_val_.end())
+    return nullptr;
+  else
+    return iter->second.get();
+}
+
+uint32_t FoldSpecConstantOpAndCompositePass::FindRecordedConst(const Const* c) {
+  auto iter = const_val_to_id_.find(c);
+  if (iter == const_val_to_id_.end())
+    return 0;
+  else
+    return iter->second;
+}
+
+bool FoldSpecConstantOpAndCompositePass::AddComponentsTo(
+    const std::vector<uint32_t>& ids, CompositeConst* composite_const) {
+  for (uint32_t id : ids) {
+    if (Const* c = FindRecordedConst(id)) {
+      composite_const->AppendComponent(c);
+    } else {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::unique_ptr<Const> FoldSpecConstantOpAndCompositePass::CreateConst(
+    const analysis::Type* type,
+    const std::vector<uint32_t>& literal_words_or_ids) {
+  std::unique_ptr<Const> new_const;
+  if (literal_words_or_ids.size() == 0) {
+    // Constant declared with OpConstantNull
+    new_const.reset(new NullConst(type));
+  } else if (type->AsBool()) {
+    assert(literal_words_or_ids.size() == 1 &&
+           "Bool constant should be delcared with one operand");
+    new_const.reset(new BoolConst(type, literal_words_or_ids.front()));
+  } else if (type->AsInteger()) {
+    new_const.reset(new IntConst(type, literal_words_or_ids));
+  } else if (type->AsFloat()) {
+    new_const.reset(new FloatConst(type, literal_words_or_ids));
+  } else if (auto* vec_type = type->AsVector()) {
+    new_const.reset(new VectorConst(vec_type, vec_type->element_type()));
+    if (!AddComponentsTo(literal_words_or_ids,
+                         new_const.get()->AsVectorConst()))
+      return nullptr;
+  } else if (type->AsStruct()) {
+    new_const.reset(new StructConst(type));
+    if (!AddComponentsTo(literal_words_or_ids,
+                         new_const.get()->AsStructConst()))
+      return nullptr;
+  } else if (type->AsArray()) {
+    new_const.reset(new ArrayConst(type));
+    if (!AddComponentsTo(literal_words_or_ids, new_const.get()->AsArrayConst()))
+      return nullptr;
+  } else {
+    return nullptr;
+  }
+  return new_const;
+}
+
+std::vector<ir::Operand> BuildOperandsFromIds(
+    const std::vector<uint32_t>& ids) {
+  std::vector<ir::Operand> operands;
+  for (uint32_t id : ids) {
+    operands.emplace_back(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                          std::initializer_list<uint32_t>{id});
+  }
+  return operands;
+}
+
+std::unique_ptr<ir::Instruction>
+FoldSpecConstantOpAndCompositePass::CreateInstruction(uint32_t id, Const* c) {
+  if (c->AsNullConst()) {
+    return std::unique_ptr<ir::Instruction>(new ir::Instruction(
+        SpvOp::SpvOpConstantNull, type_mgr_->GetId(c->type()), id, {}));
```

Comment:
We have the helper function `GetType` right? Why not using it here and in the following appearances?